Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -510,46 +517,62 @@ extension PathHierarchy {
 // MARK: Disambiguation container
 
 extension PathHierarchy {
-    /// A fixed-depth tree that stores disambiguation information and finds values based on partial disambiguation.
+    /// A container that stores values and their disambiguation information and find values based on partial disambiguation.
     struct DisambiguationContainer {
-        // Each disambiguation tree is fixed at two levels and stores a limited number of values.
-        // In practice, almost all trees store either 1, 2, or 3 elements with 1 being the most common.
+        // Each disambiguation container stores its elements in a flat list, which is very short in practice.
+        //
+        // Almost all containers store either 1, 2, or 3 elements with 1 being the most common case.
         // It's very rare to have more than 10 values and 20+ values is extremely rare.
         //
-        // Given this expected amount of data, a nested dictionary implementation performs well.
-        private(set) var storage: [String: [String: PathHierarchy.Node]] = [:]
+        // Given this expected amount of data, linear searches through an array performs well.
+        private(set) var storage = ContiguousArray<Element>()
```

Comment:
What made you decide to use an array (or a `ContiguousArray<Element>` to be precise) and not a Set? As a list of unique items, if we make `Element` conform to `Equatable`, `Set<Element>` would automatically guarantee that all the elements are unique. 
Maybe the `Element`s aren't actually equal if the hash and kind are equal but not the node?

Reply:
As a follow up question, after taking a look at `ContiguousArray`'s [documentation](https://developer.apple.com/documentation/swift/contiguousarray):

> If the array’s Element type is a struct or enumeration, `Array` and `ContiguousArray` should have similar efficiency.

What is the benefit of using `ContiguousArray` over `Array` here?

Reply:
My reason for not using a Set here is twofold:
- First, when an element already exists, the `DisambiguationContainer` needs to merge the new value with the existing value. Because of this, it's not sufficient to rely on the Set's uniqueness. Colliding elements also need to be merged. 
- Second, the `DisambiguationContainer` also needs to merge new and existing values in cases other than exact kind and hash matches. Currently this only happens when the container has a single placeholder value but in the future—when there could be more types of disambiguation—the container would need to look for existing values in more ways.

---

I _could_ implement this with a Set and still use `storage.first(where: { $0.matches(kind: kind, hash: hash) })` to check for existing matches but the Set doesn't bring any benefit over an Array in that case.

I could implement this with a Set and check for exact kind and hash matches with `Set.remove(_:)` but it would only work for the exact kind and hash match.

```swift
let element = Element(...)
if let existing = storage.remove(element) {
  existing.node.merge(with: value)
  storage.insert(existing)
} else if ...
```

I could also separate the kind and hash from the node and create a compound key

```swift
struct Key {
    let kind: String?
    let hash: String?
}
private(set) var storage = [Key: Node]()
```

This makes it very easy to look for exact matches

```swift
let key = Key(kind: kind, hash: hash)
if let existing = storage[key] {
    merge(with: value)
} else if ...
```

but when there are other types of disambiguation this breaks down and need to use `.first(where:)` to inspect the individual values again. 

Ultimately, since the purpose of this refactoring is to make it easier to store other types of disambiguation in the future (#643 already has one type of disambiguation that I want to add), I went with the solution that doesn't prioritize exact kind and hash matches.

---

I thought that there would be a bigger difference between `ContiguousArray` and `Array` but I can't measure any different except in extreme micro benchmarks. I can change it to a regular array if it makes the code easier to understand. Otherwise I don't mind keeping the `ContiguousArray` since it's a low-level implementation detail.

Reply:
For future developers it may be helpful if you could add a comment explaining why you picked a `ContiguousArray`.

Reply:
I briefly mentioned why a `Set` wouldn't help in a new comment in https://github.com/apple/swift-docc/pull/828/commits/50cc145e9db59a89a3b35b595294eb613fce4736